File: code/classes/subclass.m
#import <Foundation/Foundation.h>

// Subclass in Objective-C
@interface Person : NSObject
@property (nonatomic, strong) NSString *name;
@property (nonatomic) int age;

- (instancetype)initWithName:(NSString *)name age:(int)age;
- (NSString *)greet;
@end

@implementation Person
- (instancetype)initWithName:(NSString *)name age:(int)age {
    self = [super init];
    if (self) {
        _name = name;
        _age = age;
    }
    return self;
}

- (NSString *)greet {
    return [NSString stringWithFormat:@"Hello, I'm %@ and I'm %d years old", self.name, self.age];
}
@end

@interface Student : Person
@property (nonatomic, strong) NSString *school;

- (instancetype)initWithName:(NSString *)name age:(int)age school:(NSString *)school;
@end

@implementation Student
- (instancetype)initWithName:(NSString *)name age:(int)age school:(NSString *)school {
    self = [super initWithName:name age:age];
    if (self) {
        _school = school;
    }
    return self;
}

- (NSString *)greet {
    return [NSString stringWithFormat:@"%@ I study at %@", [super greet], self.school];
}
@end

int main() {
    Student *student = [[Student alloc] initWithName:@"Alice" age:20 school:@"MIT"];
    NSLog(@"%@", [student greet]);
    return 0;
}

File: code/classes/subclass.c
#include &lt;stdio.h&gt;
#include &lt;string.h&gt;

// Subclass in C (simulated with structs and function pointers)
struct Person {
    char name[50];
    int age;
    void (*greet)(struct Person *self);
};

void person_greet(struct Person *self) {
    printf("Hello, I'm %s and I'm %d years old\n", self-&gt;name, self-&gt;age);
}

struct Student {
    struct Person base;
    char school[50];
    void (*greet)(struct Student *self);
};

void student_greet(struct Student *self) {
    person_greet(&self-&gt;base);
    printf("I study at %s\n", self-&gt;school);
}

int main() {
    struct Student student = {{"Alice", 20, person_greet}, "MIT", student_greet};
    student.greet(&student);
    return 0;
}
